p2p/ipld/plugin/nodes/nodes.go

```diff
@@ -163,13 +144,48 @@ func DataSquareRowOrColumnRawInputParser(r io.Reader, _mhType uint64, _mhLen int
 	}
 	// to trigger the collection of nodes:
 	_ = n.Root()
-	return nodes, nil
+	return collector.Nodes(), nil
+}
+
+type NmtNodeCollector struct {
+	nodes []node.Node
+}
+
+func NewNodeCollector() *NmtNodeCollector {
+	// The extendedRowOrColumnSize is hardcode this here to avoid importing:
+	// https://github.com/lazyledger/lazyledger-core/blob/585566317e519bbb6d35d149b7e856c4c1e8657c/types/consts.go#L23
+	const extendedRowOrColumnSize = 2 * 128
+	return &NmtNodeCollector{nodes: make([]node.Node, 0, extendedRowOrColumnSize)}
+}
+
+func (n NmtNodeCollector) Nodes() []node.Node {
+	return n.nodes
+}
+
+func (n *NmtNodeCollector) Visit(hash []byte, children ...[]byte) {
+	cid := mustCidFromNamespacedSha256(hash)
+	switch len(children) {
+	case 1:
+		n.nodes = prependNode(nmtLeafNode{
+			cid:  cid,
+			Data: children[0],
+		}, n.nodes)
+	case 2:
+		n.nodes = prependNode(nmtNode{
+			cid: cid,
+			l:   children[0],
+			r:   children[1],
+		}, n.nodes)
+	default:
+		panic("expected a binary tree")
+	}
 }
 
-func prependNode(newNode node.Node, nodes *[]node.Node) {
-	*nodes = append(*nodes, node.Node(nil))
-	copy((*nodes)[1:], *nodes)
-	(*nodes)[0] = newNode
+func prependNode(newNode node.Node, nodes []node.Node) []node.Node {
+	nodes = append(nodes, node.Node(nil))
+	copy(nodes[1:], nodes)
+	nodes[0] = newNode
+	return nodes
```

Comment:
It is not the most pressing issue but I think @adlerjohn was worried about unnecessary allocations here (voiced in #152 or #155). Now that this is a struct that has the nodes as a field, couldn't this be also a method of `NmtNodeCollector` and you could modify that field directly instead of returning the modified node slice here.

Reply:
This method I believe only avoids allocations if the slice that's being prepended to never has to grow. It looks `n.nodes` is reserved with a size of `extendedRowOrColumnSize`, and should never exceed that, so there should be no allocations.

Reply:
The production code will now use a batch and immediately pass the node object the moment it is created to that batch instead of keeping track of the nodes in a slice. Not sure about the allocs in that batch object but we'd need to pass the nodes there anyway.
I'd be OK with making the NmtNodeCollector a private function again as previously and only expose the batch adder version of it.

Reply:
I made `NmtNodeCollector` a private struct https://github.com/lazyledger/lazyledger-core/pull/178/commits/c99bc38f94db16f606f4909de1b799ff23794b82, but I have kept prependNode the way it is for now. The main reason being that we are now only using it for testing purposes, and the second being that I think we should default to the idiomatic approach and optimize later. Given that a slice is [kinda a pointer](https://blog.golang.org/slices-intro) under the hood, my gut is telling me that it (weirdly) might not save on allocations or speed.

Reply:
> The main reason being that we are now only using it for testing purposes, and the second being that I think we should default to the idiomatic approach and optimize later.

Absolutely agree 💯

Reply:
> P.S. As you can see I am not satisfied with the current solution, but that is ok for now :)

Can you jot down what you dislike and what you would suggest in an issue? That would be helpful :-)